contrib/babelfishpg_tsql/src/pl_handler.c

```
@@ -4088,15 +4088,24 @@ _PG_fini(void)
  */
 
 static void
-terminate_batch(bool send_error, bool compile_error)
+terminate_batch(bool send_error, bool compile_error, int SPI_depth)
 {
 	bool		error_mapping_failed = false;
 	int			rc;
+	int 		current_spi_stack_depth;
 
 	HOLD_INTERRUPTS();
 
 	elog(DEBUG2, "TSQL TXN finish current batch, error : %d compilation error : %d", send_error, compile_error);
 
+	/*
+	 * Cleanup leftovers non internal SPI connections
+	 */
+	AtEOXact_SPI(false);
+	SPI_get_depth(&current_spi_stack_depth);
+
+	Assert(current_spi_stack_depth == SPI_depth);
```

not sure if Asserts are enabled in production or not. Better to raise a fatal error.

---

Ok